domain/personallearningplan/src/main/kotlin/uk/gov/justice/digital/hmpps/domain/personallearningplan/Exceptions.kt

```diff
@@ -27,3 +22,28 @@ class ActionPlanAlreadyExistsException(message: String) : RuntimeException(messa
  */
 class GoalNotFoundException(val prisonNumber: String, val goalReference: UUID) :
   RuntimeException("Goal with reference [$goalReference] for prisoner [$prisonNumber] not found")
+
+/**
+ * Thrown when a Goal cannot be arhived/unarchived due to the current state.
+ */
+class InvalidGoalStateException(val prisonNumber: String, val goalReference: UUID, val status: String, val action: String) :
+  RuntimeException("Could not $action goal with reference [$goalReference] for prisoner [$prisonNumber]: Goal was in state [$status] that can't be ${action}d") {
+
+  companion object {
+    const val ARCHIVE = "archive"
+    const val UNARCHIVE = "unarchive"
+  }
+}
+
+/**
+ * A general business exception that takes a string and will have a http code of 409 CONFLICT.
```

Comment:
Minor, but the way we've modelled and segregated the domain was quite deliberate and was to keep it "pure". So things like persistence and (in this case) transport concerns/terminology didn't end up "poluting" the domain.

(ie. our plan was that the domain modelled the business domain and it's behaviour, but with no regard to any given persistence or transport/api implementation. Then, at the application layer, we could catch/handle the exception and do something relevant to the transport/api . EG. in a spring-web environment we'd use a ControllerAdvice and handle it into some given response code/body. But we could also use the same domain for an SQS listener, or a console based app, or a quarkus app (if we ever moved away from spring) etc etc; and handle the exception to something relevant to the transport)

Saying `A general business exception that takes a string and will have a http code of 409 CONFLICT.` is not strictly true - the exception doesn't have the HTTP code, it's how it's handled downstream that will determine whether its a HTTP response code of some sort, or an error message in a console app, or a rejected and requeued message in an SQS app etc ...